107-tee-boot-attestation.md

```diff
@@ -0,0 +1,498 @@
+-->
+
+As CVMs are deployed on untrusted systems, it is reasonable to assume that a
+CVM user would also like to take advantage of system integrity monitoring
+provided by Keylime *in addition to* TEE technology. That is, it is reasonable
+to expect that Keylime agents will be deployed on CVMs. With that, TEE
+attestation support for environments in which Keylime will already be running
+for system integrity monitoring purposes will likely be desired.
+
+### Goals
+
+<!--
+List the specific goals of the enhancement.  What is it trying to achieve?  How will we
+know that this has succeeded?
+-->
+ * Extend Keylime Registrar to perform TEE attestation to bootstrap an agent's
+   initial vTPM state.
+ * On CVMs, extend vTPM measurements to account for TEE boot attestation within
+   Keylime.
+ * Fully support confidential computing systems' need for boot *and* runtime
+   attestation within Keylime.
+
+### Non-Goals
+
+<!--
+What is out of scope for this enhancement?  Listing non-goals helps to focus discussion
+and make progress.
+-->
+ * TODO
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation.  The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### User Stories (optional)
+
+<!--
+Detail the things that people will be able to do if this enhancement is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system.  The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+With TEE secure processors acting as the ultimate root-of-trust, much of the
+registration and attestation process's behavior should be identical to that as
+if vTPM devices were the root of trust. However, some registration steps will be
+handled before the a Keylime agent is initialized.
+
+It is critical to establish trust in a CVM as early as possible. As such,
+typical CVMs encrypt their root disks, and require a key stored on an
+attestation server to unlock the root disk and begin running any operating
+systems or applications on the CVM. To get this key from the server, a CVM would
+have to fetch its TEE evidence from the secure processor and send it as input to
+the server. If the server is able to validate the TEE evidence, it will then
+release the key. At this point, CVMs could decrypt their root disks and begin
+running their OS and applications trusting that they are in-fact running
+confidentially.
+
+[SVSM](https://github.com/coconut-svsm/svsm) is a firmware that runs in CVMs and
+is responsible for performing TEE attestation. SVSM performs TEE attestation to
+unlock the CVM's root disk and establish the initial vTPM state. SVSM will
+perform the initial communication with the Keylime Registrar.
```

Comment:
I wouldn't call it a firmware, as that can be confused with OVMF guest firmware. Typically people are referring to SVSM as a paravisor. Also SVSM isn't responsible for unlocking disks, that is a guest OS job. Thus I'd suggest rephrasing to something more like

```
[SVSM](https://github.com/coconut-svsm/svsm) is a paravisor that is the first thing
to run in a confidential VM. It can provided trusted services to the guest that are 
protected from both the host OS, and from the guest firmware and guest OS. Under
AMD SEV-SNP, SVSM will run at VMPL-0, while the guest firmware & OS will run at 
VMPL-3. One of the services SVSM can provide is a virtual TPM. By performing a
TEE attestation, SVSM can establish trust in the confidential hardware environment
and link the CVM attestation report to the virtual TPM identity, enabling the guest OS
to then trust the vTPM. 
```